### PR TITLE
feat(rush): add support for resolution only mode with pnpm

### PR DESCRIPTION
**Pull request by aramissennyeydd**
## Summary

Adding support for https://github.com/pnpm/pnpm/pull/6411. This is required for enforcing strict peer dependency validation in CI. Otherwise, pnpm will skip the check if a lockfile exists (which it generally should for install). If these issues pass CI, local developers then see issues on rush add or rush update which recreates the lockfile and reruns the resolution checks. 

## Details

Adding support for the --resolution-only flag in PNPM. This flag force pnpm to re-run dependency resolution, https://pnpm.io/cli/install#--resolution-only. Without this flag, you end up in a situation similar to https://github.com/pnpm/pnpm/issues/7087 and https://github.com/pnpm/pnpm/issues/6893. With this flag, you can enforce this in CI with an additional `rush install --resolution-only` check.

One more caveat here (that [I think is a bug with upstream PNPM](https://github.com/pnpm/pnpm/issues/8433)) is that you can't set `useFrozenLockfileForRushInstall` to true and use the `--resolution-only` flag. It needs to be false otherwise you get a version mismatch error.

## How it was tested

Tested against the main rush monorepo by removing the webpack allowed versions from `globalPeerDependencyRules` and verified that peer dependency errors do show up. Also tested against our internal rush monorepo 

## Impacted documentation

Rush CLI documentation.

### CODE REVIEW

File: common/config/rush/experiments.json [hunk 1]
```diff
@@ -9,7 +9,7 @@
    * By default, 'rush install' passes --no-prefer-frozen-lockfile to 'pnpm install'.
    * Set this option to true to pass '--frozen-lockfile' instead for faster installs.
    */
-  "usePnpmFrozenLockfileForRushInstall": true,
+  // "usePnpmFrozenLockfileForRushInstall": true,
```
Comment (iclanton) [inline common/config/rush/experiments.json:12]: Intentional change?

File: libraries/rush-lib/src/cli/actions/InstallAction.ts [hunk 2]
```diff
@@ -45,6 +46,11 @@ export class InstallAction extends BaseInstallAction {
       parameterLongName: '--check-only',
       description: `Only check the validity of the shrinkwrap file without performing an install.`
     });
+
+    this._resolutionOnlyParameter = this.defineFlagParameter({
+      parameterLongName: '--resolution-only',
+      description: `Only perform dependency resolution, useful for ensuring peer dependendencies are up to date.`
```
Comment (iclanton) [inline libraries/rush-lib/src/cli/actions/InstallAction.ts:52]:
```suggestion
      description: `Only perform dependency resolution, useful for ensuring peer dependendencies are up to date. Note that this flag is only supported when using the pnpm package manager.`
```

You can also consider only initializing this flag if pnpm is used.

**Review by iclanton — APPROVED** [review]

File: common/config/rush/experiments.json (hunk 1 above)
Comment (aramissennyeydd) [reply common/config/rush/experiments.json:12]: Nope, edited this for testing. Will revert back.

File: libraries/rush-lib/src/cli/actions/InstallAction.ts (hunk 2 above)
Comment (aramissennyeydd) [reply libraries/rush-lib/src/cli/actions/InstallAction.ts]: updated to only initialize the flag if pnpm is used in [0226e5f](https://github.com/microsoft/rushstack/pull/4893/commits/0226e5fb1e87e6c3f638b68eaa470f4f8dd7a8d1)

**Comment by iclanton** [comment]
@aramissennyeydd you'll probably need to update test snapshots

**Comment by aramissennyeydd** [comment]
@iclanton Updated, also seeing this test case `rushx should pass args to scripts` failing both locally and in CI. Taking a look.

**Comment by iclanton** [comment]
It's actually https://github.com/microsoft/rushstack/blob/01cbfb910dc5b418dd56808d89aa5b74a7d85a71/libraries/rush-lib/src/cli/test/Cli.test.ts#L16-L23 that's throwing. Looks like that code actually has a small issue that I'm fixing here: https://github.com/microsoft/rushstack/pull/4895

**Comment by aramissennyeydd** [comment]
@iclanton Interesting, thanks for taking a look into this! Just merged in master 🤞

**Comment by aramissennyeydd** [comment]
@iclanton Looks like an issue here https://github.com/aramissennyeydd/rushstack/blob/0b87d488014e00c364bea509c238d4ab2f46300f/libraries/rush-lib/src/cli/actions/BaseRushAction.ts#L116-L118, the rush configuration can be undefined. I think that's a relatively large breaking change to adjust that, what's the best path forward here?

**Comment by iclanton** [comment]
@aramissennyeydd - what you have for now is probably fine, but the type for rushConfiguration on BaseRushAction does appear to be wrong.

**Comment by iclanton** [comment]
Let's merge this in, and then I can take a look at that issue separately (or you can, if you want 🙂).

**Comment by aramissennyeydd** [comment]
@iclanton What would be the expected behavior here? If there's no rush config set, it looks like there's already a `BaseConfiglessRushAction` to handle that. I think I ran into trouble because I'm trying to access `rushConfiguration` in the constructor and its existence gets enforced in the `onExecute`.

**Comment by iclanton** [comment]
I'm not sure offhand. I'd need to poke around more in the code. I'm currently largely away from a computer for the next few days (on vacation), but I may have some time to look later in the week or this weekend.

**Pull request merged**